actionview/lib/action_view/render_parser.rb

```diff
@@ -0,0 +1,194 @@
+# frozen_string_literal: true
+
+require "action_view/ripper_ast_parser"
+
+module ActionView
+  class RenderParser # :nodoc:
+    def initialize(name, code)
+      @name = name
+      @code = code
+      @parser = RipperASTParser
+    end
+
+    def render_calls
+      render_nodes = @parser.parse_render_nodes(@code)
+
+      render_nodes.map do |method, nodes|
+        nodes.map { |n| send(:parse_render, n) }
+      end.flatten.compact
+    end
+
+    private
+      def directory
+        File.dirname(@name)
+      end
+
+      def resolve_path_directory(path)
+        if path.include?("/")
+          path
+        else
+          "#{directory}/#{path}"
+        end
+      end
+
+      # Convert
+      #   render("foo", ...)
+      # into either
+      #   render(template: "foo", ...)
+      # or
+      #   render(partial: "foo", ...)
+      def normalize_args(string, options_hash)
+        if options_hash
+          { partial: string, locals: options_hash }
+        else
+          { partial: string }
+        end
+      end
+
+      def parse_render(node)
+        node = node.argument_nodes
+
+        if (node.length == 1 || node.length == 2) && !node[0].hash?
+          if node.length == 1
+            options = normalize_args(node[0], nil)
+          elsif node.length == 2
+            options = normalize_args(node[0], node[1])
+          end
+
+          return nil unless options
+
+          parse_render_from_options(options)
+        elsif node.length == 1 && node[0].hash?
+          options = parse_hash_to_symbols(node[0])
+
+          return nil unless options
+
+          parse_render_from_options(options)
+        else
+          nil
+        end
+      end
+
+      def parse_hash(node)
+        node.hash? && node.to_hash
+      end
+
+      def parse_hash_to_symbols(node)
+        hash = parse_hash(node)
+
+        return unless hash
+
+        hash.transform_keys do |key_node|
+          key = parse_sym(key_node)
+
+          return unless key
+
+          key
+        end
+      end
+
+      ALL_KNOWN_KEYS = [:partial, :template, :layout, :formats, :locals, :object, :collection, :as, :status, :content_type, :location, :spacer_template]
+
+      RENDER_TYPE_KEYS =
+        [:partial, :template, :layout]
+
+      def parse_render_from_options(options_hash)
+        renders = []
+        keys = options_hash.keys
+
+        if (keys & RENDER_TYPE_KEYS).size < 1
+          # Must have at least one of render keys
+          return nil
+        end
+
+        unless (keys - ALL_KNOWN_KEYS).empty?
+          # de-opt in case of unknown option
+          return nil
```

Comment:
Is this superfluous after the previous check on line 99, given that `RENDER_TYPE_KEYS` is a subset of `ALL_KNOWN_KEYS`? Or am I reading it wrong?

Reply:
Ah, this logic is confusing. The previous check `if (keys & RENDER_TYPE_KEYS).size < 1` checks that you have at least one render key. This check `unless (keys - ALL_KNOWN_KEYS).empty?` checks that you don't have any keys *not* listed in ALL_KNOWN_KEYS. I am going to try `if (keys - ALL_KNOWN_KEYS).any?` since `unless _.empty?` is super backwards sounding. Thanks for catching this!

Reply:
Aaaaahhh I see! I had to read it a couple more times :sweat_smile: 

Maybe extracting some comment-methods would be useful in this case? Like:
```ruby
return nil if no_render_keys?(keys)
return nil if unknown_keys?(keys)
```